### DISCUSSION TITLE
ipython is missing the library pickleshare, how to get it installed (the right way)?

### DISCUSSION THREAD

**Discussion by bdlabitt** [Homebrew/discussions#4980] [Getting started]
### Output of `brew config`

```shell
% brew config
HOMEBREW_VERSION: 4.1.25
ORIGIN: https://github.com/Homebrew/brew
HEAD: c32bd1c7cc0c3b1b914845bddfeda53f4d877a3f
Last commit: 7 days ago
Core tap JSON: 12 Dec 20:29 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 14
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 14-core 64-bit dunno
Clang: 15.0.0 build 1500
Git: 2.43.0 => /opt/homebrew/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
% brew doctor
Your system is ready to brew.
```

### Description of issue

Simply don't know how to correctly required python libraries in a brew based system.  I thought I read pip3 install pickleshare would be a bad thing.  So what's the recommended way?  I am new to homebrew and MacOS, but do have Linux background.

More directly, the ipython formula I installed has numerous UserWarnings when run.  It seems like an incomplete installation.  Here is the output when using valid commands.

` % ipython
Python 3.12.1 (main, Dec  7 2023, 20:45:44) [Clang 15.0.0 (clang-1500.0.40.1)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.18.1 -- An enhanced Interactive Python. Type '?' for help.

/opt/homebrew/Cellar/ipython/8.18.1/libexec/lib/python3.12/site-packages/IPython/core/completerlib.py:363: UserWarning: using bookmarks requires you to install the `pickleshare` library.
  bks = self.db.get('bookmarks',{})
In [1]: cd Documents/python/home/
/opt/homebrew/Cellar/ipython/8.18.1/libexec/lib/python3.12/site-packages/IPython/core/magics/osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]
/Users/me/Documents/python/home

In [2]:      `
In my previous experience one would simply install pickleshare via pip.  What is the correct way to handle this with homebrew?  The paradigm of homebrew appears to be different, so I am asking.

Thanks.

**Comment by SMillerDev** [comment]
You can make a pull request to adjust the formula

**Reply by bdlabitt** [reply]
New to this.  Have a link that explains the process?  Something a new person could follow easily?  Don't mind doing it at all, since it makes the Homebrew better.

Also, what is the paradigm for python library management under Homebrew?  Don't want to mess up my system, if I can avoid it!

**Comment by gigaimage** [comment]
I see same UserWarning but in different context:

```
In [24]: %cd ~/Workspace/Space-PL/pythonspace/

<MYHOMEDIR>/.local/pipx/venvs/jupyter/lib/python3.12/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]
```

In [3]: !!brew config
Out[3]:
['HOMEBREW_VERSION: 4.3.7',
 'ORIGIN: https://github.com/Homebrew/brew',
 'HEAD: 43eaeca50fe3b6c755f3fc9bc6f43669e0db1039',
 'Last commit: 31 hours ago',
 'Core tap HEAD: ad0f9de36266b2698fa263f7bbac5866a6dc841e',
 'Core tap last commit: 3 days ago',
 'Core tap JSON: 24 Jun 16:31 UTC',
 'Core cask tap HEAD: c0a810852ba3d1769af156d02c94708d25fdfd6b',
 'Core cask tap last commit: 3 days ago',
 'Core cask tap JSON: 24 Jun 16:31 UTC',
 'HOMEBREW_PREFIX: /usr/local',
 'HOMEBREW_CASK_OPTS: []',
 'HOMEBREW_GITHUB_API_TOKEN: set',
 'HOMEBREW_MAKE_JOBS: 8',
 'HOMEBREW_SORBET_RUNTIME: set',
 'Homebrew Ruby: 3.3.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.3/bin/ruby',
 'CPU: octa-core 64-bit haswell',
 'Clang: 14.0.0 build 1400',
 'Git: 2.45.2 => /usr/local/bin/git',
 'Curl: 8.4.0 => /usr/bin/curl',
 'macOS: 12.7.5-x86_64',
 'CLT: 14.2.0.0.1.1668646533',
 'Xcode: 14.2']

**Comment by jdtsmith** [comment]
In the meantime, a workaround:

```python
% /opt/homebrew/Cellar/ipython/8.26.0/libexec/bin/python -m pip install pickleshare
```

**Reply by bdlabitt** [reply]
@jdtsmith the workaround is INOP.  Still get warning from ipython that pickleshare library is missing.  Yes, I used the current directory 8.29.0.  Still get the warning of the original post, which is over a year old.

**Reply by gromgit** [reply]
Not a Python expert, but I'm pretty sure:
* https://docs.brew.sh/Homebrew-and-Python#pep-668-and-virtual-environments is applicable here, and
* @jdtsmith's solution installs `pickleshare` at the _Homebrew system_ level, instead of `ipython`'s own virtualenv
* the correct solution is to run `pip install pickleshare` **at the ipython prompt** instead:
```
% ipython
Python 3.13.1 (main, Dec  3 2024, 17:59:52) [Clang 16.0.0 (clang-1600.0.26.4)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.31.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: pip install pickleshare
Collecting pickleshare
  Downloading pickleshare-0.7.5-py2.py3-none-any.whl.metadata (1.5 kB)
Downloading pickleshare-0.7.5-py2.py3-none-any.whl (6.9 kB)
Installing collected packages: pickleshare
Successfully installed pickleshare-0.7.5

[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: /opt/homebrew/Cellar/ipython/8.31.0/libexec/bin/python -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.

In [2]: import pickleshare

In [3]: 
```
